components/range_cache_memory_engine/src/range_manager.rs

```diff
@@ -1,27 +1,51 @@
 // Copyright 2024 TiKV Project Authors. Licensed under Apache-2.0.
 
 use std::{
-    collections::{BTreeMap, BTreeSet, VecDeque},
+    collections::{
+        BTreeMap,
+        Bound::{self, Excluded, Unbounded},
+    },
     result,
-    sync::Arc,
+    sync::atomic::{AtomicBool, Ordering},
 };
 
 use collections::HashMap;
-use engine_rocks::RocksSnapshot;
 use engine_traits::{CacheRange, EvictReason, FailedReason};
+use kvproto::metapb::Region;
 use tikv_util::{info, time::Instant};
 
 use crate::{metrics::observe_eviction_duration, read::RangeCacheSnapshotMeta};
 
+#[derive(PartialEq, Eq, PartialOrd, Ord, Debug, Clone, Copy, Default)]
+pub enum RegionState {
+    // waiting to be load.
+    // NOTE: in this state, the region's epoch may be older than
+    // target region in raftstore.
+    #[default]
+    Pending,
+    // Region is ready to do batch load. In this state, apply thread
+    // will directly write new kv into the skiplist.
+    ReadyToLoad,
+    // Region is handling batch loading from rocksdb snapshot.
+    Loading,
+    // Region is cached, ready to handle foreground read.
+    Active,
+    // region should be evicted, but batch loading is possible still running.
+    LoadingCanceled,
+    // region should be evicted, but there are possible active snapshot or gc task.
+    PendingEvict,
+    // evicting event is running, the region will be removed after the evcit task finished.
+    Evicting,
+}
+
 // read_ts -> ref_count
 #[derive(Default, Debug)]
 pub(crate) struct SnapshotList(pub(crate) BTreeMap<u64, u64>);
 
 impl SnapshotList {
     pub(crate) fn new_snapshot(&mut self, read_ts: u64) {
         // snapshot with this ts may be granted before
-        let count = self.0.get(&read_ts).unwrap_or(&0) + 1;
-        self.0.insert(read_ts, count);
+        *self.0.entry(read_ts).or_default() += 1;
```

Comment:
Is it possible multiple thread access function simultaneously?

Reply:
Currently no as it requires `&mut self` and it doesn't seem to be the bottleneck because this lock is very short. Maybe we can replace it with `DashMap` in the future if it becomes the bottleneck.